tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -5117,20 +5113,21 @@ class HLSLExternalSource : public ExternalSemaSource {
   /// </remarks>
   ImplicitConversionSequence TrySubscriptIndexInitialization(_In_ clang::Expr* SrcExpr, clang::QualType DestType);
 
-  void AddHLSLObjectMethodsIfNotReady(QualType qt) {
-    static_assert((sizeof(uint64_t)*8) >= _countof(g_ArBasicKindsAsTypes), "Bitmask size is too small");
-    // Everything is ready.
-    if (m_objectTypeLazyInitMask == 0)
+  void CompleteType(TagDecl *Tag) override {
+    if(Tag->isCompleteDefinition() || !isa<CXXRecordDecl>(Tag))
       return;
-    CXXRecordDecl *recordDecl = const_cast<CXXRecordDecl *>(GetRecordDeclForBuiltInOrStruct(qt->getAsCXXRecordDecl()));
+
+    CXXRecordDecl *recordDecl = cast<CXXRecordDecl>(Tag);
+    if (auto TDecl = dyn_cast<ClassTemplateSpecializationDecl>(recordDecl))
+      recordDecl = TDecl->getSpecializedTemplate()->getTemplatedDecl();
+
+    if (recordDecl->isCompleteDefinition())
+      return;
```

Comment:
Is this necessary? I think there's only one implementation of `isCompleteDefinition()`. So if it didn't bail above, it won't bail here.

Reply:
It’s a different Decl. The first call might be safe to omit because we shouldn’t get there without it being incomplete. The second call is on the template decl rather than the specialization decl. Because of the way clang recursively resolves templates the two don’t always agree.